_posts/news/2025-01-07-announcing-scalajs-1.18.0.md

```diff
@@ -0,0 +1,139 @@
+
+This is a **minor** release:
+
+* It is backward binary compatible with all earlier versions in the 1.x series: libraries compiled with 1.0.x through 1.17.x can be used with 1.18.0 without change.
+* It is *not* forward binary compatible with 1.17.x: libraries compiled with 1.18.0 cannot be used with 1.17.x or earlier.
+* It is *not* entirely backward source compatible: it is not guaranteed that a codebase will compile *as is* when upgrading from 1.17.x (in particular in the presence of `-Xfatal-warnings`).
+
+As a reminder, libraries compiled with 0.6.x cannot be used with Scala.js 1.x; they must be republished with 1.x first.
+
+## Changes with compatibility concerns
+
+### Enforce stricter rules on the use of `this` in the IR of constructors
+
+Exceptionally, **this is theoretically a binary-breaking change**.
+It is possible for previously valid IR (Intermediate Representation, as published in binaries on Maven Central) to be rejected by Scala.js 1.18.0.
+
+We now enforce, during IR checking (by default, only enabled in `fullLink` builds), that constructors abide by stricter rules.
+Before calling their `super` constructor (or delegate `this()` constructor), they may not refer to the `this` instance.
+The only exception is to *assign* a field, as in `this.foo = bar`.
+This restriction is in line with a similar rule in JVM bytecode.
+
+Unfortunately, our compiler plugin for Scala 2.x was previously generating IR that is invalid according to the new rule in one corner case situation.
+If there is a `try..catch` expression inside an argument to the `super` constructor, the compiler may lift it to a separate method.
+We compiled that method as an *instance* method, using the `this` value, instead of compiling it as a *static* method.
+
+Here is an example:
+
+{% highlight scala %}
+import scala.util.control.Breaks._
+
+class Parent(val i: Int) {
+  def this(p: Parent) = this(p.i)
+  if (i > 40) throw new IllegalArgumentException
+}
+class Child extends Parent({
+  val p = try new Parent(42) catch { case _: IllegalArgumentException => break() }
+  p
+})
+
+object Test extends App {
+  breakable(new Child)
+}
+{% endhighlight %}
+
+If your application depends on a library that contained such a code shape, you may hit the following IR checking error at link time in Scala.js 1.18.0.
+
+{% highlight none %}
+Restricted use of `this` before the super constructor call
+{% endhighlight %}
+
+If that is the case, the library needs to be recompiled with Scala.js 1.18.0 or later to be usable again.
+Make sure to [tell us on the relevant PR](https://github.com/scala-js/scala-js/pull/5019) if that happens.
+
+In order to mitigate the impact of this change, we conducted an audit of (the latest version of) all Scala.js libraries published to Maven Central.
+We did not find any library that violated the new rule.
+You may still run into the issue if you depend on a private library.
+
+### Drop support for Scala 2.12.2 to 2.12.5 and 2.13.0 to 2.13.2
```

Comment:
Did we also drop Java 16 support? (https://github.com/scala-js/scala-js/pull/5073) or did we do this earlier and is this just an artifact?

Reply:
Ah, OK. Just a misunderstanding on my part then.